File: src/main/java/com/github/Doomsdayrs/Jikan4java/core/search/TopSearch.java
package com.github.Doomsdayrs.Jikan4java.core.search;

import com.github.Doomsdayrs.Jikan4java.core.Retriever;
import com.github.Doomsdayrs.Jikan4java.enums.top.*;
import com.github.Doomsdayrs.Jikan4java.exceptions.IncompatibleEnumException;
import com.github.Doomsdayrs.Jikan4java.types.Main.Top.Top;

import java.util.concurrent.CompletableFuture;

/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 * ====================================================================
 * Jikan4java
 * 21 / 05 / 2019
 *
 * @author github.com/doomsdayrs
 */
public class TopSearch extends Retriever {
    /**
     * Searches the top charts of MAL, with all default cases
     *
     * @param tops What subType of media: anime, manga, people, characters
     * @return Top object
     */
    public CompletableFuture<Top> searchTop(Tops tops) throws IncompatibleEnumException {
        return searchCore(tops, 0, null);
    }

    /**
     * Searches the top charts of MAL, with subtype defaulted
     *
     * @param tops       What subType of media: anime, manga, people, characters
     * @param pageNumber Should be 1 by default, Each page has 50 entries; Page 2 would be the next page, so 50 - 100
     * @return Top object
     */
    public CompletableFuture<Top> searchTop(Tops tops, int pageNumber) throws IncompatibleEnumException {
        return searchCore(tops, pageNumber, null);
    }


    /**
     * Searches the top charts of MAL, page defaulted
     *
     * @param tops    What subType of media: anime, manga, people, characters
     * @param subtype Shared subcategory to core for.
     * @return Top object
     */
    public CompletableFuture<Top> searchTop(Tops tops, SharedTops subtype) throws IncompatibleEnumException {
        return searchTopSub(tops, subtype);
    }

    /**
     * Searches the top charts of MAL, page defaulted
     *
     * @param tops    What subType of media: anime, manga, people, characters
     * @param subtype The Anime subcategory to core for.
     * @return Top object
     */
    public CompletableFuture<Top> searchTop(Tops tops, AnimeTops subtype) throws IncompatibleEnumException {
        return searchTopSub(tops, subtype);
    }

    /**
     * Searches the top charts of MAL, page defaulted
     *
     * @param tops    What subType of media: anime, manga, people, characters
     * @param subtype The Manga subcategory to core for.
     * @return Top object
     */
    public CompletableFuture<Top> searchTop(Tops tops, MangaTops subtype) throws IncompatibleEnumException {
        return searchTopSub(tops, subtype);
    }

    /**
     * Searches the top charts of MAL, with subtype defaulted
     *
     * @param tops    What subType of media: anime, manga, people, characters
     * @param subtype The sub category to core for.
     * @return Top object
     */
    public CompletableFuture<Top> searchTopSub(Tops tops, TopSubType subtype) throws IncompatibleEnumException {
        return searchCore(tops, 0, subtype);
    }


    /**
     * Searches the top charts of MAL
     *
     * @param tops       What subType of media: anime, manga, people, characters
     * @param pageNumber Should be 1 by default, Each page has 50 entries; Page 2 would be the next page, so 50 - 100
     * @param subtype    The sub category to core for.
     * @return Top object
     */
    public CompletableFuture<Top> searchTop(Tops tops, int pageNumber, AnimeTops subtype) throws IncompatibleEnumException {
        return searchCore(tops, pageNumber, subtype);
    }

    /**
     * Searches the top charts of MAL
     *
     * @param tops       What subType of media: anime, manga, people, characters
     * @param pageNumber Should be 1 by default, Each page has 50 entries; Page 2 would be the next page, so 50 - 100
     * @param subtype    The sub category to core for.
     * @return Top object
     */
    public CompletableFuture<Top> searchTop(Tops tops, int pageNumber, MangaTops subtype) throws IncompatibleEnumException {
        return searchCore(tops, pageNumber, subtype);
    }

    /**
     * Searches the top charts of MAL
     *
     * @param tops       What subType of media: anime, manga, people, characters
     * @param pageNumber Should be 1 by default, Each page has 50 entries; Page 2 would be the next page, so 50 - 100
     * @param subtype    The sub category to core for.
     * @return Top object
     */
    public CompletableFuture<Top> searchTop(Tops tops, int pageNumber, SharedTops subtype) throws IncompatibleEnumException {
        return searchCore(tops, pageNumber, subtype);
    }


    /**
     * Searches the top charts of MAL
     *
     * @param tops       What subType of media: anime, manga, people, characters
     * @param pageNumber Should be 1 by default, Each page has 50 entries; Page 2 would be the next page, so 50 - 100
     * @param subtype    The sub category to core for.
     * @return Top object
     */
    public CompletableFuture<Top> searchCore(Tops tops, int pageNumber, TopSubType subtype) throws IncompatibleEnumException {
        if (tops == null) throw new EnumConstantNotPresentException(Tops.class, "Tops subType not present!");
        StringBuilder options = new StringBuilder();
        if (pageNumber == 0) pageNumber = 1;
        options.append("/");
        options.append(pageNumber);
        if (subtype != null) {
            if (tops.compatible(subtype))
                options.append("/").append(subtype);
            else throw new IncompatibleEnumException(tops + " is not compatible with " + subtype);
        }
        return retrieve(tops.getaClass(), baseURL + "/top/" + tops + options.toString());
    }


}

File: src/main/java/com/github/doomsdayrs/jikan4java/core/search/TopSearch.kt
package com.github.doomsdayrs.jikan4java.core.search

import com.github.doomsdayrs.jikan4java.common.JIKAN_URL
import com.github.doomsdayrs.jikan4java.core.JikanResult
import com.github.doomsdayrs.jikan4java.core.Retriever
import com.github.doomsdayrs.jikan4java.data.enums.top.TopSubType
import com.github.doomsdayrs.jikan4java.data.exceptions.IncompatibleEnumException
import com.github.doomsdayrs.jikan4java.data.model.main.top.base.Top
import com.github.doomsdayrs.jikan4java.data.model.main.top.base.TopListing
import com.github.doomsdayrs.jikan4java.data.model.main.top.model.anime.AnimeTop
import com.github.doomsdayrs.jikan4java.data.model.main.top.model.character.CharacterTop
import com.github.doomsdayrs.jikan4java.data.model.main.top.model.manga.MangaTop
import com.github.doomsdayrs.jikan4java.data.model.main.top.model.person.PersonTop
import java.util.concurrent.CompletableFuture

/*
 * This file is part of Jikan4java.
 *
 * Jikan4java is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Jikan4java is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Jikan4java.  If not, see <https://www.gnu.org/licenses/>.
 */

/**
 * Jikan4java
 * 21 / 05 / 2019
 *
 * @author github.com/doomsdayrs
 */
class TopSearch(
	val retriever: Retriever
) {

	/**
	 * Searches the top charts of MAL
	 *
	 * @param top       What subType of media: anime, manga, people, characters
	 * @param pageNumber Should be 1 by default, Each page has 50 entries; Page 2 would be the next page, so 50 - 100
	 * @param subtype    The sub category to core for.
	 * @return Top object
	 */
	@JvmOverloads
	inline fun <reified TOP, reified V> searchTop(
		top: TOP,
		pageNumber: Int = 0,
		subtype: TopSubType? = null
	): CompletableFuture<JikanResult<TOP>>
			where TOP : Top<V>,
			      V : TopListing {
		val options =
			StringBuilder("/${if (pageNumber == 0) 1 else pageNumber}")
		if (subtype != null) {
			if (top.compatible(subtype)) options.append("/").append(subtype)
			else throw IncompatibleEnumException("$top is not compatible with $subtype")
		}
		println("TOP IS: ${TOP::class.simpleName}")
		println("var IS: ${top::class.simpleName}")

		return retriever("$JIKAN_URL/top/${top.name}$options")
	}

	companion object {
		val PERSON by lazy { PersonTop() }
		val MANGA by lazy { MangaTop() }
		val CHARACTER by lazy { CharacterTop() }
		val ANIME by lazy { AnimeTop() }
	}
}